### ISSUE TITLE
Auto-created releases don't include all artifacts

### ISSUE THREAD

**Issue by tillig** [stoplightio/spectral#1767]
In the last month or so it appears some work has been going on :tada: but as part of the CI/CD process, `stoplight-bot` is creating a release for each tag 😦  - this [`@stoplight/spectral-ruleset-migrator-v1.3.0` release](https://github.com/stoplightio/spectral/releases/tag/%40stoplight%2Fspectral-ruleset-migrator-v1.3.0) is an example.

These releases **don't include the executables or other artifacts** like in a [fully versioned release](https://github.com/stoplightio/spectral/releases/tag/v6.0.0-alpha2) - it's just a source snapshot. While this isn't a showstopper, it really breaks the interface for https://github.com/stoplightio/spectral/releases/latest and https://api.github.com/repos/stoplightio/spectral/releases/latest, which scripts and build tasks can use to download the latest release of the tooling.

Is there a way to either...

- Stop releasing these `@stoplight/*` tags (and delete the existing incomplete "releases" so they don't show up anymore)? OR
- Make sure these releases also include all of the executables and everything so the `latest` release actually has a complete latest release?

**Comment by P0lip** [comment]
Hey!

> stoplight-bot is creating a release for each tag :frowning: - this @stoplight/spectral-ruleset-migrator-v1.3.0 release is an example.

Yup, that's absolutely intended. We want an independent release tag for each package, in order to generate a changelog.

> These releases don't include the executables or other artifacts like in a fully versioned release - it's just a source snapshot. While this isn't a showstopper, it really breaks the interface for https://github.com/stoplightio/spectral/releases/latest and https://api.github.com/repos/stoplightio/spectral/releases/latest, which scripts and build tasks can use to download the latest release of the tooling.

While no final decision has been made yet, it's possible we won't be including binaries from now on.
This makes little sense, now that we have different versioning.
Let me expand that topic.

`@stoplight/cli` will be released less frequently now, as rulesets, functions, core, etc. are all located outside of that package.
In other words, the CLI package is the code that provides CLI capabilities, thus releases happen only when we change, add, remove a flag, or add a new command etc.

What we'll certainly bring back is a working install.sh script, but one can still use `npm` or `yarn` (which always have been the recommended ways of installing Spectral).
However, even with that updated install.sh, one may still have to install the rulesets package separately if they intend to use the OAS/AsyncAPI rulesets.

**Comment by tillig** [comment]
I suppose from one standpoint that makes sense - having release tags per package - but from a different standpoint... I mean, I'm not sure how other folks consume this, but I'm guessing quite a few do it through the CLI. This definitely makes it harder to see the overall set of changes from one "version of the tool" to the next. Though I recognize it's not "one thing" that constitutes the tool, it's somewhat different based on who's considered the consumer and how they consume. I can't say seeing a release per package is, at least for me and my limited set of use cases, super helpful. Maybe I'm in the minority, and that's OK.

From an "integrate this tool into a build pipeline" (Azure DevOps, GitHub Actions) **having a standalone binary tool was really valuable**. It meant not having to worry about Node and `npm` issues, etc. It "just worked." (Great example - right now we're fighting with `bash` exiting with a non-zero exit code when `npm` runs and notifies you of an update. Makes Docker packaging of tools sort of painful.) However, it sounds like we're going to need to switch to `npm` for installation, which means changing up build scripts and custom build tasks - not a showstopper, but definitely some unplanned work, to be sure.

Might be worth updating [the documentation](https://meta.stoplight.io/docs/spectral/ZG9jOjYyMDc0Mw-installation) to add a note that `install.sh` isn't working and maybe remove the reference to standalone executables if they're going away. That way new folks coming on won't start down that direction if it's disappearing. It might also be worth explicitly calling out that `npm` or `yarn` are the _recommended_ ways to do the installation - the current docs read like all the ways listed (standalone executable, Docker) are equally valid and supported.

**Comment by P0lip** [comment]
> I suppose from one standpoint that makes sense - having release tags per package - but from a different standpoint... I mean, I'm not sure how other folks consume this, but I'm guessing quite a few do it through the CLI. 

You can still consume Spectral this way, nothing has changed in this regard.
The only caveat is that the only working way to accomplish this is to install it via `npm` or `yarn` as stated in the README.
I'll make sure to get install.sh sorted out.

> This definitely makes it harder to see the overall set of changes from one "version of the tool" to the next. 

Although I understand this concern, having CLI including everything was a real blocker for us.
This implied that any time we introduced a breaking change elsewhere, say in rulesets or functions, we got to release a breaking CLI release. Moreover, folks couldn't quite stick to a particular version of a ruleset, meaning they'd need to constantly look at CHANGELOG and keep updating their rulesets.
After the change we're more aligned with other linters like, for example, ESLint. 

> From an "integrate this tool into a build pipeline" (Azure DevOps, GitHub Actions) having a standalone binary tool was really valuable. It meant not having to worry about Node and npm issues, etc. It "just worked." (Great example - right now we're fighting with bash exiting with a non-zero exit code when npm runs and notifies you of an update. Makes Docker packaging of tools sort of painful.) However, it sounds like we're going to need to switch to npm for installation, which means changing up build scripts and custom build tasks - not a showstopper, but definitely some unplanned work, to be sure.

Unless you need to upgrade ASAP, you might want to stick around with v5 for a short while.
As mentioned, we'll certainly try to find a solution to meet that use case.

One of the solutions that has come to my mind is to release weekly binaries, that albeit still not linked to a release, could be easily installed using `install.sh`.
Each such a "release" would come up with the latest set of changes from all packages.

Does that make sense / help with your use case?

**Comment by tillig** [comment]
> install it via npm or yarn as stated in the README

The README also directs you to the docs that say you can use `install.sh`, download the binaries, or use a Docker distribution, which is why I mentioned it would be good to get the docs updated. The breaking releases/latest thing affects all those places.

<img width="419" alt="README screen shot" src="https://user-images.githubusercontent.com/1156571/128715629-1e5ab923-88ee-4bd8-bd32-22c4d6f11f56.png">

> After the change we're more aligned with other linters like, for example, ESLint.

So, I get why you're doing it and totally respect it. The actual linter is made up of several components that can be independently released. However, it might be good to consider that the CLI isn't "a package in-and-of-itself" but more of a meta-package designed for distribution.

Spectral lints JSON and YAML, right? Those are really more markup than programmatic; while JSON is "JavaScript Object Notation, it isn't actually _JavaScript_ or _TypeScript_ per se and can be used in any project type.

I am likely preaching to the choir, but let me use an analogy for my own benefit. Let's say we were still doing web services and I created an XML linter that could work on any XSD schema but had special rules for WSDL. I wrote the linter in C#/.NET because that's what my other projects are in and it made sense. I compiled cross-platform binaries and distributed the linter three ways:

1. As a binary you can take and use.
2. As [a `dotnet` global tool](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools) - if you have the full .NET SDK installed, you can choose to install that way.
3. As a Docker container that basically has the `dotnet` global tool in it.

For folks on Java projects or Node projects, likely the best route for them is that binary because they probably don't have the .NET SDK installed. On build servers, the binary is the clear way to go because, again, the build servers don't necessarily have .NET SDK installed (and if they do, maybe it's not the right runtime) and pulling the Docker container is far slower than just grabbing the executable.

Now I want to start versioning the rules in my linter faster. You don't get the CLI binary anymore, and likely not the Docker container either since they have the same problem - they come with a pre-packaged version of the rules. Now in your Java or Node or Ruby project, to use the linter you'll need to:

- Have the .NET SDK installed.
- Add a [.NET `global.json`](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json?tabs=netcore3x) to the root of your project to indicate the .NET SDK being used.
- Add a .NET `dotnet-tools.json` to the root of your project to indicate the linter version being used.

Your non-.NET project now has a bunch of .NET stuff in it and .NET dependencies so you can consume the linter that lints... code that isn't .NET.

**I do not intend to be argumentative or combative.** Honestly, I love Spectral and we use it _all the time_ and that's actually how I encountered this - the inability to use the standard GitHub releases/latest to get the latest CLI.

It does worry me a bit that ESLint is the example linter used as the way to go - ESLint lints JavaScript and is a JavaScript dependency, sort of like how [Roslyn analyzers lint C# code and are .NET dependencies](https://docs.microsoft.com/en-us/visualstudio/code-quality/roslyn-analyzers-overview?view=vs-2019). I feel like if you want the updated rules and you're using the CLI... you should get a new version of the CLI with the updated rules baked in. The CLI is that metapackage with everything in it.

**I think the weekly releases with a metapackage semantic version would work** as suggested. However, I'd potentially take that a step further and let the GitHub releases/latest _also work_ because, again, that's a programmatic interface to access that metapackage. I don't think it's safe to assume folks are going to be able to run `install.sh` on their Windows machines to get the latest Windows binary. There are also some security issues with `curl | bash` piping that make it more palatable to locate and download the executable directly if possible.

Finally, **I'm curious how this decision affects the VS Code integration**. That is, the VS Code integration is a static package similar to the CLI. The way ESLint in VS Code works, it doesn't respect global tools; it assumes you have a local Node project and a local ESLint reference in the project, then it uses that. It makes it easy for ESLint because you can use any version with any set of rules... but it definitely makes that assumption you're in a Node project. Which, for ESLint, is a reasonable assumption. Not so much with a .NET microservice.

Anyway, honestly, **thank you very much for your patience and time discussing this with me**. I thought it would be valuable to offer a perspective here and note the surprise that releases/latest has stopped working; I hope I haven't been too frustrating.

For now we'll go through our pipelines and scripts to stick on the latest v5 release instead of looking for the latest binary. We'll wait to see how this all settles out as far as whether there are binaries; how to locate them if there are; and so on. Once that's known, we can update our pipelines and scripts accordingly. If there won't be binaries that would be good to know sooner rather than later because we'll need to modify our build agent images to also include Node and possibly work through the process of getting the Spectral packages periodically scanned and added to our internal Nexus Repository for `npm` access.

**Comment by P0lip** [comment]
> I am likely preaching to the choir, but let me use an analogy for my own benefit. Let's say we were still doing web services and I created an XML linter that could work on any XSD schema but had special rules for WSDL. I wrote the linter in C#/.NET because that's what my other projects are in and it made sense. I compiled cross-platform binaries and distributed the linter three ways:

> As a binary you can take and use.
> As a dotnet global tool - if you have the full .NET SDK installed, you can choose to install that way.
> As a Docker container that basically has the dotnet global tool in it.
> For folks on Java projects or Node projects, likely the best route for them is that binary because they probably don't have the .NET SDK installed. On build servers, the binary is the clear way to go because, again, the build servers don't necessarily have .NET SDK installed (and if they do, maybe it's not the right runtime) and pulling the Docker container is far slower than just grabbing the executable.
> 
> Now I want to start versioning the rules in my linter faster. You don't get the CLI binary anymore, and likely not the Docker container either since they have the same problem - they come with a pre-packaged version of the rules. Now in your Java or Node or Ruby project, to use the linter you'll need to:
> 
> Have the .NET SDK installed.
> Add a .NET global.json to the root of your project to indicate the .NET SDK being used.
> Add a .NET dotnet-tools.json to the root of your project to indicate the linter version being used.

I see what you mean here, but fortunately you don't need to keep anything in your project, as everything is installed globally and lives somewhere in the folder maintained by npm that's outside of the project dir (depending on the platform)
I do realize however that the biggest pain here is the fact one has to install Node.js together with npm, and that there's no solution for it than installing Node.js, which obviously makes things a bit more complicated, in particular on Windows.


> Finally, I'm curious how this decision affects the VS Code integration. That is, the VS Code integration is a static package similar to the CLI. The way ESLint in VS Code works, it doesn't respect global tools; it assumes you have a local Node project and a local ESLint reference in the project, then it uses that

I'm not a VS Code user, but WebStorm lets you override the ESLint package you want to execute. You can pick any file. I'd be surprised if VS Code didn't have the same option. 
We are yet to update the VS Code integration, but the existing one should be fully functional - it's just it still runs the older version.

> I hope I haven't been too frustrating.

Not at all! Thanks for the feedback and the discussion.

> Anyway, honestly, thank you very much for your patience and time discussing this with me. I thought it would be valuable to offer a perspective here and note the surprise that releases/latest has stopped working;

Thanks for your detailed write-up and the time spent, I frankly appreciate it.
Believe me or not, I truly understand your troubles, and will do my best to resolve them.

> For now we'll go through our pipelines and scripts to stick on the latest v5 release instead of looking for the latest binary. We'll wait to see how this all settles out as far as whether there are binaries; how to locate them if there are; and so on. Once that's known, we can update our pipelines and scripts accordingly. If there won't be binaries that would be good to know sooner rather than later because we'll need to modify our build agent images to also include Node and possibly work through the process of getting the Spectral packages periodically scanned and added to our internal Nexus Repository for npm access.

There will be binaries, don't worry about it.
I'm just unsure whether each release will have them included or whether it's a weekly scheduled release, but the following reasoning...

>  I think the weekly releases with a metapackage semantic version would work as suggested. However, I'd potentially take that a step further and let the GitHub releases/latest also work because, again, that's a programmatic interface to access that metapackage.

...convinced me that it might be not such a bad idea to simply include binaries to each package.

We'll see. Either way, I'll ensure to make the process is as smooth and convenient as possible.
I'll keep you updated, you can expect to hear from me this week in this issue.

LMK if you have any other feedback or any other concern.

**Comment by tillig** [comment]
I think this sounds reasonable. Basically, as long as I can get binaries my build servers will be happy. 😄 

> WebStorm lets you override the ESLint package

They sort of took this to the extreme [in the VS Code / ESLint integration where you _have_ to override the ESLint package](https://github.com/microsoft/vscode-eslint/blob/77b70686d5b7709e657c18de4aa278bac72ac7e7/server/src/eslintServer.ts#L885). Which is to say, you aren't allowed to install it as a global tool, you _must_ have it as part of your TypeScript/JavaScript project. Which, of course, means you also _must_ have `packages.json` and so on.

If the VS Code integration follows that pattern, it'd be a similar challenge to having to install the CLI using `npm` into your current project.

Granted, if it's a global tool it's less of an issue on a per-project basis, but then you have the same problem ESLint ran into where different projects may want different versions of the linter and that's not so easy with a global tool.

Anyway, thanks again. I look forward to seeing how this all turns out!

**Comment by P0lip** [comment]
@tillig the latest release has all the artifacts included. 
The subsequent ones will also have them attached, so hopefully it's all good.
My apologies for any inconvenience caused.

**Comment by tillig** [comment]
🎉 YAY! I'll unpin and assume `latest` will work. I think that should solve the `install.sh` issues, too.

**Comment by P0lip** [comment]
Let me know how it goes! :crossed_fingers:

**Comment by tillig** [comment]
We'll have to do a bit of work since [Azure DevOps pipeline tasks assume installed tools have a release in semantic versioning format](https://github.com/microsoft/azure-pipelines-tool-lib/blob/b0924f6b3445ec162c1db7cfd89c547b171fa0a4/tool.ts#L298) like either `v1.2.3` or `1.2.3` so when the release is `@stoplight/spectral-ruleset-migrator-v1.4.0` it blows up not understanding what the version is.

I also noticed the `spectral --version` now shows the version of the CLI package (I assume) like `1.2.1` so it appears _older_ than any other releases. That could be a challenge for determining what the latest installed or cached version is from a programmatic perspective.

**Comment by tillig** [comment]
[I've added an issue to the Azure DevOps task library repo](https://github.com/microsoft/azure-pipelines-tool-lib/issues/128) to see if they can allow non-semver versions. It looks like the corresponding GitHub Actions library has it already set up for non-semver versions (too bad I'm not on GitHub Actions).

**Comment by P0lip** [comment]
> We'll have to do a bit of work since Azure DevOps pipeline tasks assume installed tools have a release in semantic versioning format like either v1.2.3 or 1.2.3 so when the release is @stoplight/spectral-ruleset-migrator-v1.4.0 it blows up not understanding what the version is.

Ah, hm. That's unfortunate. Does it only support semantic versioning or are other versioning systems supported too?
I was thinking about including the date, similar to some of other vendors do it, but I'm not sure if that'd help in your case.

> I also noticed the spectral --version now shows the version of the CLI package (I assume) like 1.2.1 so it appears older than any other releases. That could be a challenge for determining what the latest installed or cached version is from a programmatic perspective.

Yeah, it's the version of the current stoplight/spectral-cli package.

**Comment by tillig** [comment]
It really only supports semantic versioning - `x.y.x`. It [uses the `semver-compare` package](https://www.npmjs.com/package/semver-compare) in its [version location code](https://github.com/microsoft/azure-pipelines-tool-lib/blob/b0924f6b3445ec162c1db7cfd89c547b171fa0a4/tool.ts#L92) to help locate the latest if you specify a version range.

I'm guessing the issue I filed won't meet with success because looking at that code I linked, they also support ranges like `1.x` to be specified in the tool downloader and that, too, assumes the release is a semantic version.

Now, you could use the date if you formatted it like `yyyy.m.d` (no leading zeroes) because `2021.8.15` would be older than `2021.8.16` and that follows the three-segment semantic version rules. But, again, it can't just be present in the tool (like `spectral --version`), it would actually have to be the release name in GitHub because that's what the tool is looking at: Use GitHub API to get the list of releases, use the name of the release as the version. That's really the crux of the issue - having the `@not/a-semantic-version-v1.2.3` format "versions" for the release breaks the ability to determine the version from the release name - as soon as one of these non-semver-releases is the latest, it all blows up.

**Comment by P0lip** [comment]
> It really only supports semantic versioning - x.y.x. It uses the semver-compare package in its version location code to help locate the latest if you specify a version range.

Ah, I see, too bad.

Is there any particular reason you rely on that package to install Spectral? I suppose it might come up with some caching, but given the fact we have a fixed release cadence (we have a cron job that's executed daily, and if there's any commit to be published, a new release is cut), implementing caching manually might be actually feasible and, hopefully, not too complex. 
Moreover, since we always distribute dedicated binaries for Windows, one doesn't need to do any extractions nor anything else, and as far as I see that tool also accounts for such a possibility.
I'm not sure if Azure offers any cron jobs and how each job is provisioned, but perhaps, given the above assumptions, it's possible to set up it manually?

LMK if there's anything I could help with. I really feel bad that the changes we introduced caused issues in your workflow 😞

**Comment by tillig** [comment]
> Is there any particular reason you rely on that package to install Spectral?

Well... yeah. The task library the Azure DevOps (and GitHub Actions) team provides handles a lot of stuff for you, like:

- Making sure you don't double-download the same tool in the same pipeline
- Putting the tool on your path so it's usable
- Caching the tool in an appropriate location so you don't stomp any other tools and don't have to write it into your source checkout location
- Dealing with automatic unzipping of tools shipped in zip packages
- Reading and parsing the GitHub API to find the right release based on specific version, version spec, or `latest` (which _used to work_ here)

...and so on. Creating a custom build task from scratch to do all that didn't seem like there was a lot of value since everything was already written and would behave consistently with the existing built-in tool downloaders Microsoft provides (`kubectl`, `helm`, etc.) just by using that library.

The build agents in Azure DevOps and GitHub Actions are dynamically provisioned, not something one can configure. Any custom tools you want in the pipeline that aren't [provided by one of the virtual environments](https://github.com/actions/virtual-environments) needs to be downloaded and installed on the fly each build. You don't get your own custom build agent image without work and licensing cost. VM spins up, runs your build, VM spins down. Only the build is in our control.

The issues as I see them are:

- **This is a monorepo, not one-repo-per-package**: Spectral as a whole is made up of several packages which each want their own release schedule. I can't find an analogous type of repo where this happens.
    - [The .NET runtime](https://github.com/dotnet/runtime) is a monorepo with several packages, but they all release at one time (and largely with the same version) which definitely has its pros and cons; and the assets aren't downloaded from GitHub.
    - [eslint](https://github.com/eslint/eslint) isn't a monorepo.
    - [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/releases) is close, provides plugins to eslint, but it has one "master version" and increments it regardless of which library changed; then the library which changed is noted in the release notes rather than being the name of the release itself.
    - [babel](https://github.com/babel/babel) is also close, and provides eslint plugins, but also just has one "master version" and increments that regardless of what changed.
    - [airbnb](https://github.com/airbnb/javascript) has plugins for eslint, is a monorepo, but doesn't provide releases at all - it's `npm` or bust. (Which makes sense for a thing only ever consumed by Node projects anyway.)
    - Tools like [Azure CLI `az`](https://github.com/Azure/azure-cli) provide installation via other means, not GitHub.
- **Some consumers want a _tool_, some consumers want the _libraries_**: I don't consume Spectral as a set of independent libraries, I see it as a single tool. Other folks do see it as a set of independent libraries and consume them as such.

The breaker came when the view of "what a release is" went from "it's a tool" to "it's a set of libraries." The strategy here of trying to do both is, at least from my experience, really uncommon and will be harder for machine-based systems (like Azure DevOps, GitHub Actions, etc.) to automate or standardize any sort of interaction with.

If I could make a recommendation, it'd be to follow what typescript-eslint or babel are doing and go back to "it's a tool." If libraries change, the release notes for that version of the tool can contain subsections for each library change/version, but maybe don't have a per-library release + name. Consumers of the tool _as a tool_ don't care about the individual library changes; consumers of the libraries won't get those updates from GitHub releases, they'll get the updates from `npm` or similar.

**Comment by P0lip** [comment]
Thanks for the detailed write-up!

Speaking of the tool, it appears like it indeeds simplifies the process, thus it's probably best to stick to it as you noted.

> This is a monorepo, not one-repo-per-package: Spectral as a whole is made up of several packages which each want their own release schedule. I can't find an analogous type of repo where this happens.

I'd argue Spectral is considerably different from all of these packages.
It's kind of like you had _both_ ESLint and `typescript-eslint` and the airbnb config in the same repo.
In other words, it's the engine itself and a number of configs/plugins. 
While I'm aware that ESLint comes with their own set of rules, they rarely change, and as a result they don't hamper the release process.

The problem we tried to fix with the independent versioning (which, by the wya, isn't that unpopular after all - even [Lerna](https://github.com/lerna/lerna#independent-mode) supports it out of the box) is to ensure we can proceed quicker with rulesets changes. As mentioned earlier, we often couldn't release a new breaking change, since this would result in a new major release of a new package.
This was particularly painful in Studio, where people couldn't choose the version of Spectral they run.
In other words, we had to be very cautious with releases we make, and that eventually led to situations where certain PRs containing breaking changes were actually released after months.
Having independent versioning helps greatly here, as people can use whatever version they want to.

> If I could make a recommendation, it'd be to follow what typescript-eslint or babel are doing and go back to "it's a tool." If libraries change, the release notes for that version of the tool can contain subsections for each library change/version, but maybe don't have a per-library release + name. Consumers of the tool as a tool don't care about the individual library changes; consumers of the libraries won't get those updates from GitHub releases, they'll get the updates from npm or similar.

Point taken, albeit I'm not certain whether we'd change the versioning.

That said, I'm open in providing another mean of distribution.
Overall, I don't think that GitHub releases is the best way to handle updates - quite a few npm packages don't use GH releases at all, since `npm` is the source of truth for the Node.js ecosystem (and Spectral generally belongs to it), thus if there's any other repository we could distribute binaries to, I'm all ears!
I feel like this would be the best solution here.

**Comment by tillig** [comment]
> I don't think that GitHub releases is the best way to handle updates - quite a few npm packages don't use GH releases at all, since npm is the source of truth for the Node.js ecosystem

I think that's the trouble with having a tool and libraries - GitHub releases _is_ a great solution to handle updates... for the _binaries_ or the standalone tool.

Now, the other option you have is to start supporting other installation mechanisms. For example...

- Homebrew for MacOS
- `apt-get` and other package managers for Linux
- `winget` and Chocolatey for Windows

I think GH releases is probably the simpler way to go.

Perhaps splitting the CLI part and binaries off to a separate repo and releasing from there? This repo could still have the per-library release schedule and the binaries/CLI could be, say, periodic or whatever like ESLint does it.

Interesting example, Lerna - looks like [their releases are also semantic versions](https://github.com/lerna/lerna/releases/tag/v4.0.0). (I get it, the build system supports this sorta weird release thing. But I can't find any tools or libraries that I've encountered yet that do this with GitHub releases.)

**Comment by tillig** [comment]
Out of curiosity - you mentioned `npm` is the source of truth for Node.js updates and many packages don't even use GH releases. It seems like that might be a reason to dedicate the GH releases to the tool part of things.

Who is the intended consumer of these non-semver releases? Is it expected library consumers would come in there to see changes? Something else? Could you generate a GitHub Pages markdown page on commit with the changelog rather than issuing a release?

**Comment by P0lip** [comment]
> I think that's the trouble with having a tool and libraries - GitHub releases is a great solution to handle updates... for the binaries or the standalone tool.

True, I agree with you.

> Perhaps splitting the CLI part and binaries off to a separate repo and releasing from there? This repo could still have the per-library release schedule and the binaries/CLI could be, say, periodic or whatever like ESLint does it.

I'm afraid this won't be possible for other reasons than engineering-related.

> Out of curiosity - you mentioned npm is the source of truth for Node.js updates and many packages don't even use GH releases. It seems like that might be a reason to dedicate the GH releases to the tool part of things.
> Who is the intended consumer of these non-semver releases? Is it expected library consumers would come in there to see changes? Something else? Could you generate a GitHub Pages markdown page on commit with the changelog rather than issuing a release?

I really like this idea. Honestly, we used to maintain our own changelog.md anyway, therefore it wouldn't change much if we had these automated changelogs or not.
I believe I'll go down this path - not going to happen today or tomorrow, but hopefully next week assuming I won't need to rewrite the whole release script.

**Comment by tillig** [comment]
You are _royalty among people_. Thanks - this will make life a lot easier.

**Comment by P0lip** [comment]
Heyo!
Just wanted to follow-up - I'm going to get this done tomorrow.
I've done some research around semantic-release, and I'll need to make a few minor changes in our config, but we should be good.

**Comment by P0lip** [comment]
@tillig as promised, it's out https://github.com/stoplightio/spectral/releases/tag/v6.0.0
I also moved changelogs around and explained how versioning works - this work is happening [here](https://github.com/stoplightio/spectral/tree/chore/changelogs).

**Comment by tillig** [comment]
Sweet! I will unpin in our build and give it a shot on Monday!

**Comment by P0lip** [comment]
I'm going to close this one, but shall you have any further troubles, reach out to me here, please.
I believe we should be good, as this works exactly as it used to do in the past, so I don't expect any change in the behavior, but do let me know in case something is still off for you.

**Comment by tillig** [comment]
Verified this works, thanks!

Would have checked yesterday but... 😩  let's just say "Windows domain account" + Mac = _pain_. That's 11 hours of password reset and account synchronization I'll never get back.

But this works! So yay!

**Comment by P0lip** [comment]
> Would have checked yesterday but... 😩 let's just say "Windows domain account" + Mac = pain. That's 11 hours of password reset and account synchronization I'll never get back.


Haha, no worries.

Glad to hear it works!
